usr.bin/script/script.c

```diff
@@ -0,0 +1,481 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/time.h>
+#include <sys/param.h>
+#include <sys/uio.h>
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <paths.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <time.h>
+#include <tzfile.h>
+#include <unistd.h>
+#include <util.h>
+
+#define	DEF_BUF	65536
+
+struct stamp {
+	uint64_t scr_len;	/* amount of data */
+	uint64_t scr_sec;	/* time it arrived in seconds... */
+	uint32_t scr_usec;	/* ...and microseconds */
+	uint32_t scr_direction;	/* 'i', 'o', etc (also indicates endianness) */
+};
+
+static FILE	*fscript;
+static int	master, slave;
+static int	child, subchild;
+static size_t	outcc;
+static int	usesleep, rawout;
+static int	quiet, flush;
+static const char *fname;
+
+static int	isterm;
+static struct	termios tt;
+
+static void	done(void);
+static void	dooutput(void);
+static void	doshell(const char *);
+static void	fail(void);
+static void	finish(int);
+static void	scriptflush(int);
+static void	record(FILE *, char *, size_t, int);
+static void	consume(FILE *, off_t, char *, int);
+static void	childwait(void);
+static void	playback(FILE *);
+
+int
+main(int argc, char *argv[])
+{
+	ssize_t scc;
+	size_t cc;
+	struct termios rtt;
+	struct winsize win;
+	int aflg, pflg, ch;
+	char ibuf[BUFSIZ];
+	const char *command;
+
+	aflg = 0;
+	pflg = 0;
+	usesleep = 1;
+	rawout = 0;
+	quiet = 0;
+	flush = 0;
+	command = NULL;
+	while ((ch = getopt(argc, argv, "ac:dfpqr")) != -1)
+		switch(ch) {
+		case 'a':
+			aflg = 1;
+			break;
+		case 'c':
+			command = optarg;
+			break;
+		case 'd':
+			usesleep = 0;
+			break;
+		case 'f':
+			flush = 1;
+			break;
+		case 'p':
+			pflg = 1;
+			break;
+		case 'q':
+			quiet = 1;
+			break;
+		case 'r':
+			rawout = 1;
+			break;
+		case '?':
+		default:
+			(void)fprintf(stderr,
+			    "Usage: %s [-c <command>][-adfpqr] [file]\n",
+			    getprogname());
+			exit(EXIT_FAILURE);
+		}
+	argc -= optind;
+	argv += optind;
+
+	if (argc > 0)
+		fname = argv[0];
+	else
+		fname = "typescript";
+
+	if ((fscript = fopen(fname, pflg ? "r" : aflg ? "a" : "w")) == NULL)
+		err(EXIT_FAILURE, "fopen %s", fname);
+
+	if (pflg)
+		playback(fscript);
+
+	if (tcgetattr(STDIN_FILENO, &tt) == -1 ||
+	    ioctl(STDIN_FILENO, TIOCGWINSZ, &win) == -1) {
+		if (errno != ENOTTY) /* For debugger. */
+			err(EXIT_FAILURE, "tcgetattr/ioctl");
+		if (openpty(&master, &slave, NULL, NULL, NULL) == -1)
+			err(EXIT_FAILURE, "openpty");
+	} else {
+		if (openpty(&master, &slave, NULL, &tt, &win) == -1)
+			err(EXIT_FAILURE, "openpty");
+		isterm = 1;
+	}
+
+  if (!quiet){
+    (void)printf("Script started, output file is %s\n", fname);
+    fflush(stdout);
```

Comment:
Is it really required to introduce `fflush` here?

Reply:
It was necessary a couple of weeks ago, but it no longer is. I've removed it.